### PR TITLE
Deduplication buffer spooling

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeManager.java [hunk 1]
```diff
@@ -58,9 +58,10 @@
      * A new sink instance is created by the coordinator for every task attempt (see {@link Exchange#instantiateSink(ExchangeSinkHandle, int)})
      *
      * @param handle returned by {@link Exchange#instantiateSink(ExchangeSinkHandle, int)}
+     * @param preserveRecordsOrder preserve order of records within a partition
```
Comment (losipiuk) [inline core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeManager.java:61]: maybe add a note that it may impact performance.

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientFactory.java [hunk 2]
```diff
@@ -123,7 +125,12 @@ public ThreadPoolExecutorMBean getExecutor()
     }
 
     @Override
-    public DirectExchangeClient get(LocalMemoryContext systemMemoryContext, TaskFailureListener taskFailureListener, RetryPolicy retryPolicy)
+    public DirectExchangeClient get(
+            QueryId queryId,
+            ExchangeId exchangeId,
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientFactory.java:140]: change of this interface should go to last commit. New args are not used in `Identify exchanges by ExchangeId`

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java [hunk 3]
```diff
@@ -35,6 +35,7 @@
     private int clientThreads = 25;
     private int pageBufferClientMaxCallbackThreads = 25;
     private boolean acknowledgePages = true;
+    private DataSize deduplicationBufferSize = DataSize.of(32, Unit.MEGABYTE);
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:38]: I am not convinced we need this one. Why can't we have just `maxBufferSize` and set it to different value if we decide to enable QUERY-based retries and to different value without retries?

And even if we want specific property I would not use `deuplication` in it. Feels to low-level.
Maybe `exchange.query-retries-max-buffer-size`?

File: testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java [hunk 4]
```diff
@@ -28,6 +28,8 @@ private FaultTolerantExecutionConnectorTestHelper() {}
                 .put("query.initial-hash-partitions", "5")
                 .put("fault-tolerant-execution-target-task-input-size", "10MB")
                 .put("fault-tolerant-execution-target-task-split-count", "4")
+                // to trigger spilling
+                .put("exchange.deduplication-buffer-size", "1kB")
```
Comment (losipiuk) [inline testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java:32]: can we have some test (just single use-case) to verify that spilling actually happens if this value is low, and does not if it is high?

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java (hunk 3 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:38]: The deduplication buffer is used on coordinator and we may want to configure it differently. For example for pipelined execution we may want to have a higher (or lower) buffer size.

File: testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java (hunk 4 above)
Comment (arhimondr) [reply testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java:32]: I tried to cover it with unit tests. I may try to extract statistics for the deduplication buffer and check whether spilledBytes is not 0

Comment (arhimondr) [reply testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java:32]: It's actually not that easy to extract operator statistics. There's a unit test that verifies whether spilling actually takes places. Verifying it here is more of verify whether the configuration parameter has been correctly passed to the buffer, as it is extremely unlikely that the hundreds of queries executed are not actually exceeding the 1kB limit.   I wonder if it's worth it?

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 5]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+                sinkInstanceHandle = exchange.instantiateSink(sinkHandle, 0);
+                exchange.noMoreSinks();
+                exchangeSink = exchangeManager.createSink(sinkInstanceHandle, true);
+
+                writeBuffer = new DynamicSliceOutput(DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+            }
+
+            if (!pageBuffer.isEmpty()) {
+                for (Map.Entry<TaskId, List<Slice>> entry : asMap(pageBuffer).entrySet()) {
+                    writeToSink(entry.getKey(), entry.getValue());
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:533]: In theory we can keep them in buffer and don't write to sink. Writing them saves memory and simplifies logic, keeping them will incur reduce the amount of data written to the sink.

File: core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java [hunk 6]
```diff
@@ -316,7 +332,12 @@ public void addSplit(RemoteSplit split)
                 ExchangeDataSource dataSource = delegate.get();
                 if (dataSource == null) {
                     if (split.getExchangeInput() instanceof DirectExchangeInput) {
-                        DirectExchangeClient client = directExchangeClientSupplier.get(systemMemoryContext, taskFailureListener, retryPolicy);
+                        DirectExchangeClient client = directExchangeClientSupplier.get(
+                                taskId.getQueryId(),
+                                new ExchangeId(format("direct-exchange-%s-%s", taskId.getStageId().getId(), sourceId)),
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java:337]: Why do we need to include `sourceId` here? If I remember correctly, there will be only one deduplication buffer on the coordinator?

File: core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java [hunk 7]
```diff
@@ -147,154 +176,421 @@ public void testIsBlocked()
...
                         .build(),
                 ImmutableMap.of(),
-                ImmutableList.of("p0a0v0"));
+                DataSize.of(1, KILOBYTE),
+                1,
+                ImmutableList.of(createPage("p0a0v0", DataSize.of(2, KILOBYTE))));
+
+        // discard single page, with no spilling
+        testPollPages(
+                RetryPolicy.QUERY,
```
Comment (linzebing) [inline core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java:403]:
```suggestion
                RetryPolicy.TASK,
```

File: core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java [hunk 8]
```diff
@@ -147,154 +176,421 @@ public void testIsBlocked()
...
+                        .build(),
+                ImmutableMap.of(),
+                DataSize.of(10, KILOBYTE),
+                0,
+                ImmutableList.of(
+                        createPage("p0a1v0", DataSize.of(3, KILOBYTE))));
+
+        // discard single page, with spilling
+        testPollPages(
+                RetryPolicy.QUERY,
```
Comment (linzebing) [inline core/trino-main/src/test/java/io/trino/operator/TestDeduplicatingDirectExchangeBuffer.java:416]:
```suggestion
                RetryPolicy.TASK,
```

File: testing/trino-testing/src/main/java/io/trino/testing/BaseFailureRecoveryTest.java [hunk 9]
```diff
@@ -106,6 +106,8 @@ protected final QueryRunner createQueryRunner()
                         .put("failure-injection.request-timeout", new Duration(REQUEST_TIMEOUT.toMillis() * 2, MILLISECONDS).toString())
                         .put("exchange.http-client.idle-timeout", REQUEST_TIMEOUT.toString())
                         .put("query.initial-hash-partitions", "5")
+                        // to trigger spilling
+                        .put("exchange.deduplication-buffer-size", "1kB")
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseFailureRecoveryTest.java:110]: I don't know why, but kilo bytes are actually lower case: https://github.com/airlift/units/blob/0e46afc2cf1196da4cb5d772f6b07482b648c860/src/main/java/io/airlift/units/DataSize.java#L302

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 10]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+                    }
+                    throw e;
+                }
+
+                exchange = exchangeManager.createExchange(new ExchangeContext(queryId, exchangeId), 1);
+
+                ExchangeSinkHandle sinkHandle = exchange.addSink(0);
+                sinkInstanceHandle = exchange.instantiateSink(sinkHandle, 0);
+                exchange.noMoreSinks();
+                exchangeSink = exchangeManager.createSink(sinkInstanceHandle, true);
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:526]: The `preserveRecordsOrder` flags tells the `ExchangeSink` to preserve order in which records are added to a partition. Thus for example writing a single partition into multiple files in a randomized fashion should not be allowed. It doesn't impact the current (local file system) implementation as the records are appended to a single file in order. For S3 implementation I guess it should be the same.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 11]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+            bufferedPageCount += pages.size();
+        }
+
+        private static long getRetainedSizeInBytes(List<Slice> pages)
+        {
+            long result = 0;
+            for (Slice page : pages) {
+                result += page.getRetainedSize();
+            }
+            return result;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:549]: I copied it from the streaming buffer where this operation is done using a loop (probably for performance reasons)? It shouldn't matter that much though, but at the same time it doesn't change readability much either.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 5 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:533]: The sink is likely to have a buffer on it's own. Maintaining two buffers will increase the memory footprint and make the implementation more complex.

File: core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java (hunk 6 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java:337]: Because a single stage may be consuming more than a single source

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 12]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+            closed = true;
+
+            pageBuffer.clear();
+            pageBufferRetainedSizeInBytes = 0;
+            bufferedPageCount = 0;
+            writeBuffer = null;
+            try (Closer closer = Closer.create()) {
+                closer.register(exchange);
+                if (exchangeSink != null) {
+                    closer.register(exchangeSink::abort);
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:660]: Good catch. I forgot to nullify it when creating an output. The idea is "when it's non null - it means it's active and it has to be aborted, when it's null - it's either hasn't been created or already closed"

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 13]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+                }
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    @NotThreadSafe
+    private interface OutputSource
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:670]: Read it more as of `This is a source of the buffer output`

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 14]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+        @Override
+        public Slice getNext()
+        {
+            if (finished) {
+                return null;
+            }
+            if (exchangeSource == null) {
+                CompletableFuture<List<ExchangeSourceHandle>> sourceHandlesFuture = exchange.getSourceHandles();
+                if (!sourceHandlesFuture.isDone()) {
+                    return null;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:759]: The order of invokation of `isBlocked` and `poolPage` is inherently racy. It is possible that the `isBlocked` return "not blocked" (because of the `outputReady` feature) but then it become blocked on the `sourceHandlesFuture`.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 15]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+                int stageId = buffer.getInt(0);
+                int partitionId = buffer.getInt(Integer.BYTES);
+                int attemptId = buffer.getInt(Integer.BYTES * 2);
+                TaskId taskId = new TaskId(new StageId(queryId, stageId), partitionId, attemptId);
+                if (!selectedTasks.contains(taskId)) {
+                    continue;
+                }
+                return buffer.slice(Integer.BYTES * 3, buffer.length() - Integer.BYTES * 3);
+            }
+            close();
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:781]: It is to have only the `finished` variable instead of both `finished` and `closed` that seems redundant. It set's `finished` to `true` and closes the source. It is also correct to set `finished` to true here directly, but then the source should be closed in `close` and be controlled by a second `closed` variable.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 16]
```diff
@@ -385,29 +405,427 @@ public synchronized void close()
...
+            if (exchangeSource != null) {
+                CompletableFuture<?> blocked = exchangeSource.isBlocked();
+                if (!blocked.isDone()) {
+                    return nonCancellationPropagating(asVoid(toListenableFuture(blocked)));
+                }
+            }
+            CompletableFuture<List<ExchangeSourceHandle>> sourceHandles = exchange.getSourceHandles();
+            if (!sourceHandles.isDone()) {
+                return nonCancellationPropagating(asVoid(toListenableFuture(sourceHandles)));
+            }
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:806]: The `if (exchangeSource != null) {` implies that the source is already created, thus the `sourceHandles` are already available

File: core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java (hunk 6 above)
Comment (linzebing) [reply core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java:337]: Does it hold for stages not involving coordinators? I don't see you include `sourceId` there.

Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java:337]: This is somehow confusing to be honest. The external exchanges are named after the producing stage, that is always unique. Direct exchanges named after a consuming stage (that may consume data from multiple producing stages). Naming a direct exchange after a producing stage is not possible, as a single consumer may consume data from multiple source stages.

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java (hunk 3 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:38]: Yeah true - both are used in same query - makes sense

File: testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java (hunk 4 above)
Comment (losipiuk) [reply testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java:32]: Not pushing too much. But getting stats should not be that hard with `AbstractTestQueryFramework#assertQueryStats`

File: testing/trino-testing/src/main/java/io/trino/testing/BaseFailureRecoveryTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (losipiuk) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseFailureRecoveryTest.java:110]: I guess that is just the way it is: https://en.wikipedia.org/wiki/Metric_prefix

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java [hunk 17]
```diff
@@ -140,4 +141,16 @@ public DirectExchangeClientConfig setAcknowledgePages(boolean acknowledgePages)
         this.acknowledgePages = acknowledgePages;
         return this;
     }
+
+    public DataSize getDeduplicationBufferSize()
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:145]: @NotNull?

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java [hunk 18]
```diff
@@ -140,4 +141,16 @@ public DirectExchangeClientConfig setAcknowledgePages(boolean acknowledgePages)
...
+    {
+        return deduplicationBufferSize;
+    }
+
+    @Config("exchange.deduplication-buffer-size")
+    public DirectExchangeClientConfig setDeduplicationBufferSize(DataSize deduplicationBufferSize)
+    {
+        this.deduplicationBufferSize = deduplicationBufferSize;
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:153]: do we need `requireNotNull` here? I am not sure. It is not common, but I don't know if this is an omission or not.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 19]
```diff
@@ -153,30 +189,25 @@ public synchronized void addPages(TaskId taskId, List<Slice> pages)
...
-        checkState(!successfulTasks.contains(taskId), "task is finished: %s", taskId);
-        checkState(!failedTasks.containsKey(taskId), "task is failed: %s", taskId);
-
         if (failure != null) {
             return;
         }
 
+        checkState(allTasks.contains(taskId), "task is not registered: %s", taskId);
+        checkState(!successfulTasks.contains(taskId), "task is finished: %s", taskId);
+        checkState(!failedTasks.containsKey(taskId), "task is failed: %s", taskId);
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:198]: move of this code block looks unrelated. Separate commit?

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 20]
```diff
@@ -217,26 +248,20 @@ public synchronized void noMoreTasks()
...
 
-    private synchronized void checkInputFinished()
+    private void checkInputFinished()
     {
         if (failure != null) {
             return;
         }
 
-        if (inputFinished) {
+        if (outputSource != null) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:257]: check `outputReady` future instead? This would be more consistent with other checks in class.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 21]
```diff
@@ -385,29 +405,428 @@ public synchronized void close()
...
+                verify(exchange == null, "exchange is not expected to be initialized");
+                verify(sinkInstanceHandle == null, "sinkInstanceHandle is not expected to be initialized");
+                verify(writeBuffer == null, "writeBuffer is not expected to be initialized");
+
+                try {
+                    exchangeManager = exchangeManagerRegistry.getExchangeManager();
+                }
+                catch (TrinoException e) {
+                    if (EXCHANGE_MANAGER_NOT_CONFIGURED.toErrorCode().equals(e.getErrorCode())) {
+                        throw new TrinoException(EXCHANGE_MANAGER_NOT_CONFIGURED, "Exchange manager must be configured for failure recovery capabilities to be fully functional");
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:516]: Why do you need to rethrow changing message. Is the original one poor - maybe we should just change it?

Also it would be nice to verify configuration earlier that when we really need to use the spooling exchange. Otherwise we may not notice configuration problem for very long time until we hit the affected codepath.
Actually, I believe it is being validated in some other place, but then it would be nice to add a comment that this is the case.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 22]
```diff
@@ -385,29 +405,428 @@ public synchronized void close()
...
+                pageBuffer.clear();
+                pageBufferRetainedSizeInBytes = 0;
+            }
+
+            writeToSink(taskId, pages);
+            bufferedPageCount += pages.size();
+        }
+
+        private static long getRetainedSizeInBytes(List<Slice> pages)
+        {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:544]: Should you also account `exchangeSink.getMemoryUsage()` here?

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 23]
```diff
@@ -385,29 +405,428 @@ public synchronized void close()
...
+            return result;
+        }
+
+        private void writeToSink(TaskId taskId, List<Slice> pages)
+        {
+            verify(exchangeSink != null, "exchangeSink is expected to be initialized");
+            verify(writeBuffer != null, "writeBuffer is expected to be initialized");
+            for (Slice page : pages) {
+                // wait for the sink to unblock
+                getUnchecked(exchangeSink.isBlocked());
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:558]: Do we need to explicitly wait here. What is the semantics of `exchangeSink.add` if sink is blocked?

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 22 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:544]: never mind. This is for given list of pages.

**Review by losipiuk — APPROVED** [review]
Nice

File: testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java (hunk 4 above)
Comment (arhimondr) [reply testing/trino-testing/src/main/java/io/trino/testing/FaultTolerantExecutionConnectorTestHelper.java:32]: The stats about whether spilling occur will be stored in the `OperatorInfo` within the `OperatorStats`. To get the right `OperatorStats` you somehow need to know the stageId, pipelineId and the operatorId for the `ExchangeOperator` that uses `DeduplicationBuffer`. While it is possible to figure (or even hardcode) this is not the only challenge.

For `SELECT` queries the `DeduplicationBuffer` is used in the `ExchangeClient` created to fetch final query results (in `Query.java`) which is not attached to any operator. The stats from this buffer are not available at all. 

So it leaves only the `INSERT` queries to play with. But for `INSERT` queries the data size produced by table writers is not well predictable. It may vary based on seemingly unrelated changes to the connector. 

In theory it is possible to set the limit to something extremely low, e.g.: `10B`. But then it will cause all queries to spill, and what I wanted to achieve here is to exercise both spilling and non spilling scenarios.

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java (hunk 17 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:145]: Good catch

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java (hunk 18 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientConfig.java:153]: For stats we usually prefer validation by `@NotNull` that I forgot to add :-)

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 19 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:198]: Actually yeah, let me move it

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 20 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:257]: Future is unblocked asynchronously =\

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 21 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:516]: The default message just says "Exchange manager is not configured". The idea was to provide a message that is more actionable and easier to understand. 

As an afterthought the alternative message is not very actionable either. I first was thinking about changing the message to:

> Capacity of the deduplication buffer required for failure recovery to function has been exceeded. Please configure exchange manager to allow the buffer to spill to an external storage.

But then I thought that the message of the error thrown by the `ExchangeManager` should be more actionable in general. 

I'm going to remove this overwrite of a message and improve the message in the `ExchangeManager` as a separate commit.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 23 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:558]: Due to inherent racy nature I would assume that the explicit wait can be omitted (e.g.: a Sink can technically become blocked in between the `isBlocked` and `add`). However it still feels like it is better to follow the contract.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 21 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:516]: >  Otherwise we may not notice configuration problem for very long time until we hit the affected codepath.

I'm not sure if we should always try to eagerly disallow failure recovery when the ExchangeManager is not configured. For example query level retries may function properly if the buffer capacity is never exceeded. For task level retries some queries (for example simple selects) may function properly as well.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java]: For task-level retries the set of working queries is so limited that I strongly feel we should check at the very beginning that exchange manager is configured.

For query-level retries ... maybe not. Yet the cost of configuring it so it uses local disk is small. And then it is easier for Trino users to know that when they configured a cluster and it works for some test queries, it will actually work for other ones too. 
On the other hand it makes life harder for people who run Trino with disk-less machines.

**Comment by losipiuk** [comment]
@arhimondr let me know if there is still sth to be addressed or good to go

**Comment by arhimondr** [comment]
Nothing from my side. But let's wait for @linzebing to take another look.

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (linzebing) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:750]: Sorry I don't understand here, the sink only has one partition and we called `addSink`, `noMoreSinks` as well as `sinkFinished` before this point, then `sourceHandlesFuture` must have finished.

[the remark this answers is not in the document]
Comment (linzebing) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:750]: I don't fully understand. When `isBlocked` returns "not blocked", then `sourceHandlesFuture` must have been unblocked, otherwise `isBlocked` won't return "not blocked". 

When `outputReady` is done, you return `outputSource.isBlocked()`

[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:750]: The contract between `getNext()` and `isBlocked()` invocation order is weak. The implementation usually shouldn't assume that the `isBlocked` is always in a not blocked state when the `getNext` is invoked.

The contract is weak as the engine is allowed to call `isBlocked` and `getNext` from multiple threads. When more than a single thread calls `isBlocked` and receives a signal that the  buffer is not blocked both threads may decide to call `getNext` but it is possible that the buffer will have data only for one of them. 

Under these circumstances it is hard to rely on strong guarantees. For example engine may sometime allow "spurious" wakeups and call `getNext` even when `isBlocked` hasn't been resolved. Thus it doesn't feel safe to assert here.

**Review by linzebing — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java [hunk 24]
```diff
@@ -66,37 +90,56 @@
...
+        if (failure != null || closed) {
+            return immediateVoidFuture();
+        }
+
+        if (!outputReady.isDone()) {
+            return nonCancellationPropagating(outputReady);
+        }
+
+        checkState(outputSource != null, "outputSource is expected to be set");
+        return outputSource.isBlocked();
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:142]:
```suggestion
         return nonCancellationProgating(Futures.transformAsync(outputRead, ignored -> {
           synchronized (this) {
               if (outputSource != null) {
                   return outputSource.isBlocked();
               }
               return immediateVoidFuture();
           }
         }, directExecutor()));
```

Comment (linzebing) [reply core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java:142]: Similarly for `ExchangeOutputSource::isBlocked()`. Thoughts?
